src/hotspot/share/jfr/periodic/sampling/jfrCPUTimeThreadSampler.cpp

```diff
@@ -0,0 +1,763 @@
+    lost_samples = new_lost_samples;
+  }
+  return lost_samples;
+}
+
+void JfrCPUTimeTraceQueue::ensure_capacity(u4 capacity) {
+  if (capacity != _capacity) {
+    set_capacity(capacity);
+  }
+}
+
+void JfrCPUTimeTraceQueue::ensure_capacity_for_period(u4 period_millis) {
+  u4 capacity = CPU_TIME_QUEUE_CAPACITY;
+  if (period_millis > 0 && period_millis < 10) {
+    capacity = (u4) ((double) capacity * 10 / period_millis);
+  }
+  ensure_capacity(capacity);
+}
+
+void JfrCPUTimeTraceQueue::clear() {
+  Atomic::release_store(&_head, (u4)0);
+}
+
+static int64_t compute_sampling_period(double rate) {
+  if (rate == 0) {
+    return 0;
+  }
+  return os::active_processor_count() * 1000000000.0 / rate;
+}
+
+class JfrCPUTimeThreadSampler : public NonJavaThread {
+  friend class JfrCPUTimeThreadSampling;
+ private:
+  Semaphore _sample;
+  NonJavaThread* _sampler_thread;
+  double _rate;
+  bool _autoadapt;
+  volatile int64_t _current_sampling_period_ns;
+  volatile bool _disenrolled;
+  volatile bool _stop_signals;
+  volatile int _active_signal_handlers;
+  volatile bool _is_async_processing_of_cpu_time_jfr_requests_triggered;
+
+  JfrCPUTimeThreadSampler(double rate, bool autoadapt);
+
+  void start_thread();
+
+  void enroll();
+  void disenroll();
+  void update_all_thread_timers();
+
+  void autoadapt_period_if_needed();
+
+  void set_rate(double rate, bool autoadapt);
+  int64_t get_sampling_period() const { return Atomic::load(&_current_sampling_period_ns); };
+
+  void sample_thread(JfrSampleRequest& request, void* ucontext, JavaThread* jt, JfrThreadLocal* tl);
+
+  // sample all threads that are in native state (and requested to be sampled)
+  void stackwalk_threads_in_native();
+
+  void stackwalk_thread_in_native(JavaThread* thread);
+
+protected:
+  virtual void post_run();
+public:
+  virtual const char* name() const { return "JFR CPU Time Thread Sampler"; }
+  virtual const char* type_name() const { return "JfrCPUTimeThreadSampler"; }
+  void run();
+  void on_javathread_create(JavaThread* thread);
+  bool create_timer_for_thread(JavaThread* thread, timer_t &timerid);
+  void on_javathread_terminate(JavaThread* thread);
+
+  void handle_timer_signal(siginfo_t* info, void* context);
+  void init_timers();
+  void stop_timer();
+
+  void trigger_async_processing_of_cpu_time_jfr_requests();
+};
+
+
+JfrCPUTimeThreadSampler::JfrCPUTimeThreadSampler(double rate, bool autoadapt) :
+  _sample(),
+  _sampler_thread(nullptr),
+  _rate(rate),
+  _autoadapt(autoadapt),
+  _current_sampling_period_ns(compute_sampling_period(rate)),
+  _disenrolled(true),
+  _stop_signals(false),
+  _active_signal_handlers(0),
+  _is_async_processing_of_cpu_time_jfr_requests_triggered(false) {
+  assert(rate >= 0, "invariant");
+}
+
+void JfrCPUTimeThreadSampler::trigger_async_processing_of_cpu_time_jfr_requests() {
+  Atomic::release_store(&_is_async_processing_of_cpu_time_jfr_requests_triggered, true);
```

Comment:
What prior stores are you ensuring should be visible by using release semantics here?